crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
+            .chars()
+            .take_while(|c| c.is_whitespace())
+            .any(|c| c != ' ');
+
+        if tab_or_non_ascii_space {
+            // We strip the indentation that is shared with the docstring
+            // statement, unless a line was indented less than the docstring
+            // statement, in which we strip only this much indentation to
+            // implicitly pad all lines by the difference, or all lines were
+            // overindented, in which case we strip the additional whitespace
+            // (see example in [`format_docstring`] doc comment). We then
+            // prepend the in-docstring indentation to the string.
+            let indent_len = indentation_length(trim_end) - self.stripped_indentation_length;
+            let in_docstring_indent = " ".repeat(usize::from(indent_len)) + trim_end.trim_start();
+            text(&in_docstring_indent, Some(line.offset)).fmt(self.f)?;
         } else {
-            empty_line().fmt(f)
+            // Take the string with the trailing whitespace removed, then also
+            // skip the leading whitespace.
+            let trimmed_line_range = TextRange::at(line.offset, trim_end.text_len())
+                .add_start(self.stripped_indentation_length);
+            if self.already_normalized {
+                source_text_slice(trimmed_line_range).fmt(self.f)?;
+            } else {
+                // All indents are ascii spaces, so the slicing is correct.
+                text(
+                    &trim_end[usize::from(self.stripped_indentation_length)..],
+                    Some(trimmed_line_range.start()),
+                )
+                .fmt(self.f)?;
+            }
+        }
+
+        // We handled the case that the closing quotes are on their own line
+        // above (the last line is empty except for whitespace). If they are on
+        // the same line as content, we don't insert a line break.
+        if !line.is_last {
+            hard_line_break().fmt(self.f)?;
+        }
+
+        Ok(())
+    }
+
+    /// Given a sequence of lines from a code snippet, format them and return
+    /// the formatted code as a sequence of owned docstring lines.
+    ///
+    /// This routine generally only returns an error when the recursive call
+    /// to the formatter itself returns a `FormatError`. In all other cases
+    /// (for example, if the code snippet is invalid Python or even if the
+    /// resulting reformatted code snippet is invalid Python), then `Ok(None)`
+    /// is returned. In this case, callers should assume that a reformatted
+    /// code snippet is unavailable and bail out of trying to format it.
+    ///
+    /// Currently, when the above cases happen and `Ok(None)` is returned, the
+    /// routine is silent about it. So from the user's perspective, this will
+    /// fail silently. Ideally, this would at least emit a warning message,
+    /// but at time of writing, it wasn't clear to me how to best do that.
```

Comment:
Hmm yeah. I guess the warning utilities only exist in the linter. We could move those macros to (e.g.) `ruff_macro`, or something, but I'm not hugely concerned about this.

Reply:
All righty. I'll keep it out of scope for now. cc @zanieb since I think you might care about this and might have other ideas  or opinions.

Reply:
Issue created: https://github.com/astral-sh/ruff/issues/8856